HomeBuilder - Initial contribution

kubawolanin opened the pull request:
Here's my attempt to port [generator-openhab](https://github.com/kubawolanin/generator-openhab) functionality to the openHAB UI.

Closes #178

Discussion that led me to this can be found [here](https://community.openhab.org/t/openhab-generator-design-your-smart-home-in-no-time/).

~~#### *Very early stage*. ***Not for merging***~~
~~The code is ugly and doesn't work currently - submitting a PR for visibility.~~
~~Note: "openHAB Home Designer" title is temporary.~~

### Preview:
Here's a little demo for you to play around: **[home generator](http://kubawolanin.com/home-test/)**
It doesn't reflect the TODO list underneath. It's temporarily hosted on my server if that's ok. Will remove it after it's all merged.

![generator](https://user-images.githubusercontent.com/2270505/28747074-0e92bf00-7497-11e7-9cb4-2f1d984acc04.gif)

### TODO:

- [x] i18n
- [x] Styling + layout
- [x] Swap Bootstrap 4 stylesheet with [Material](http://daemonite.github.io/material/)
- [ ] Syntax highlight with [highlight.js](https://github.com/isagalaev/highlight.js) ([for vue](https://github.com/metachris/vue-highlightjs))
- [ ] Ability to dynamically add new rooms (per @ThomDietrich's request)
- [x] Quick 'Copy' button
- [x] REST API request construction for non-textual config ( `PUT /items` - requires https://github.com/eclipse/smarthome/issues/4060 to be implemented first)
- [x] Update documentation and README
- [x] Move changes to dashobard bundle to another PR
- [x] Create a separate bundle
- [x] Get rid of plural `Group`-items names in favor of `gSomething` (meh...)
- [x] Make the dynamic form produce real home structure
- [x] Sort out templating (from EJS to Vue templates)
- [x] Create a webpack-based build chain
- [x] Move all dependencies to `package.json`
- [x] Dynamic column formatting for Items (ascii-table)
- [x] Items preview
- [x] Sitemap preview
- [x] Ability to toggle `{ channel="" }` suffixes for non-Group items
- [x] Ability to toggle icons
- [x] `checklists` replaced with [vue-multiselect](http://monterail.github.io/vue-multiselect/)
- [x] Better Sitemaps design (a frame per floor with the room groups)
- [x] HABPanel JSON generator
- [x] Add optional HomeKit/HueEmulation/Alexa tags (see [Item tagging](http://docs.openhab.org/addons/io/homekit/readme.html#item-configuration))

And stuff that I  will need someone's help with:
- [ ] Saving the `*.items` and `*.sitemap` files inside `$openhab-config`
- ~Saving items directly in the JSONDb~ I'll use REST API
- [ ] Saving HABPanel dashboards directly (related to https://github.com/openhab/org.openhab.ui.habpanel/issues/209)
- ~Making this page actually accessible from somewhere ;-)~ I'll create a bundle

Pinging: @kaikreuzer @ThomDietrich @rkoshak @ghys

Signed-off-by: Kuba Wolanin <hi@kubawolanin.com> (github: kubawolanin)

kaikreuzer commented:
@kubawolanin That already looks awesome, thank you so much for giving in wrt porting it to the UI! I hope you also agree that this way it is more user friendly and will reach a much wider audience 😀

bkumio commented:
@kubawolanin , Really like that approach. Just one other thought I like to have as feature to create it with some Multilanguage capabilities. So for German, instead of rooms it would be “Räume” ,  Ground floor  would be “Erdgeschoss” and Bathroom would be "Badezimmer” ,etc….

What about including the translations in a resource file and make the wizard let you choose the language by a flag symbol then creating the structure in the corresponding language? 

Happy to help with German/French translations and testing 😊

kubawolanin commented:
Hey @bkumio absolutely!
This exact feature is included in my TODO list above under enigmatic ` i18n` bullet point ;-)

I'm about to publish `en.json` (English) and `pl.json` (Polish) files to get us started. Hopefully more people will join and share their translations as well.
Cheers!

bkumio commented:
ah cool , learnt something about i18n today... (not a developer)....

Just some thinking during lunchtime…..
couple of questions (translations almost done)
- The first questions should be translated too, or? Your english file has got the questions but the PL language file doesn’t?
- I realized that for some words for singular and plural do have the same word in German. Does this create a problem if we generate two items with the potential same name?
- Any visualisation problems with using the German Umlauts "ä/ö/ü/ ?
- Shutters I think we need another one called "Blinds" as these are two different things, or?
- Perhaps worth adding a “kids bath(room)” ? 
- Perhaps worth adding a “garden”
- Perhaps worth adding a “penthouse”
…. The are even more 😊

- Plus another feature request would be to create Items already prepared for KNX syntax . In the sense of we could have the items already with a "{knx = "0/0/0"} for a switch or "{ knx = "0/0/0+<1/1/1" }" incl. the status input GA. Devil is in the detail for the shutters and other stuff
- How do you want to handle if I have more than one light in a room. Perhaps when clicking/activating a room/item is says “(1)” . Another click says “(2)” another right-click subtracts -1 . Then when creating the items you will add nn-number of it…. Anyway the finetune needs to be done manually but the basic structure is there…. which I like

ThomDietrich commented:
Hey @kubawolanin @bkumio localization is important, sure thing. I was wondering if this setting will only be used for the generator or if you will also transfer it to the global configuration of the openHAB configuration? Sure should be ;)

kaikreuzer commented:
> if this setting will only be used for the generator

Please note that we already [have such a setting](https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/i18n/LocaleProvider.java) (which imho should be used by the generator), which can be set in the Paper UI under Configuration->System->Regional Settings.
The plan is to also derive a default setting from the user location. The problem might be that this is only set by the dashboard, but not already by the setup wizard; I am wondering, if the generator might be best placed on the dashboard as well (which would allow entering it even after the initial setup is done)?

ThomDietrich commented:
Exactly what I was pointing at ;)

ThomDietrich commented:
Also please do not derive a language from the location of the user. That's an insensitive move if you think about it. The language selection should be one of the first settings the user can change (as @kubawolanin already does in his gif above) and a default should at best be derived from the browser language.

kaikreuzer commented:
> Also please do not derive a language from the location of the user. That's a rather insensitive move if you think about it.

You are right - as the language is available in the browser, this is indeed the much better choice.

@htreu How hard would it be to enhance the dashboard location setting feature by a check, if a language is set in config admin already and if not, automatically set the one from the browser?

kubawolanin commented:
> @htreu How hard would it be to enhance the dashboard location setting feature by a check, if a language is set in config admin already and if not, automatically set the one from the browser?

Hey @kaikreuzer I've actually figured it out myself :) In the upcoming commit you'll see something like this in [geolocation.js](https://github.com/openhab/openhab-core/blob/master/bundles/org.openhab.ui.dashboard/web/js/geolocation.js):

```javascript

        $.getJSON(window.location.origin + '/rest/services/org.eclipse.smarthome.core.i18nprovider/config', function(response) {
            window.language = response;

            if (!response.location) {
                $('.geolocation').show()
                output.html("<div class=\"spinner spinner--steps\"><img src=\"img/spinner.svg\"></div>")
                navigator.geolocation.getCurrentPosition(success, error)
            }
        })
```

BTW. So far I'm extending the `org.openhab.ui.dashboard` bundle - I'm reusing e.g. Bootstrap, jQuery and overall styling/layout structure from it. 
Would it make sense to create a completely new bundle just to place this generator in `/start/index` page? Or rather continue within this bundle and make it accessible in a different way? 

Cheers,
Kuba

kubawolanin commented on bundles/org.openhab.ui.dashboard/web/js/geolocation.js:
```diff
@@ -35,7 +37,7 @@
     function send(configuration) {
         $.ajax({
-            url : '../rest/services/org.eclipse.smarthome.core.i18nprovider/config',
+            url : window.location.origin + '/rest/services/org.eclipse.smarthome.core.i18nprovider/config',
```
Relative path wouldn't let me test it from `http://openhab:8080/static/` page. This approach is safer. Let me know if there are any obstacles.

htreu replied:
You may also change line 27 `$.getJSON('../rest/....` accordingly.

htreu replied:
ah, sorry. your example above already has this. nice!

kubawolanin commented on bundles/org.openhab.ui.dashboard/web/js/home-designer.js:
```diff
@@ -302,4 +460,9 @@ var vue = {
+
+// TODO: Use a Promise instead
+$(document).ready(() => {
+    vm.setLanguage();
```
This doesn't work yet as the REST response comes asynchronously (as it should). I'm gonna use Promise instead of waiting for document to be loaded.
Maybe I'll try not to use jQuery at all, since this is the only place I'm using `$`.

kubawolanin commented on bundles/org.openhab.ui.dashboard/web/js/home-designer.js:
```diff
@@ -60,23 +60,23 @@ var rooms = [
 ];
 
 var devices = [
```
I'll probably move these definitions to a separate file.

kubawolanin commented:
Alright!
Here's a little demo for you to play around: **[home generator](http://kubawolanin.com/home-test/)**
It's temporarily hosted on my server if that's ok. Will remove it after it's all merged.

### Remarks
- Still no language support
- No plural names for device groups yet
- No HABPanel generator yet
- There's a bug with the floors number input that I need to handle
- Need to work on the order of rooms and devices selection - right now it feels random as I'm simply pushing new fields to the array.

### What's done
- `Items` file is now aligned dynamically thanks to lightweight [ascii-table](https://github.com/sorensen/ascii-table) - @ThomDietrich you might like this 😉 
- If there's just one floor, Items have no prefix (e.g. `GF_Kitchen`, `FF_Bathroom` etc).
- If there's more than one floor, Sitemap generates them instead of single rooms. I'm not sure if it's right, but worth discussing. 

```xtend
sitemap our_home label="Our Home" {
    Frame {
        Group item=GF
        Group item=FF
    }
    Frame {
        Group item=Temperature
        Group item=Fan
        Group item=Window
        Group item=Shutter
        Group item=Humidity
    }
}
```

ThomDietrich commented:
Hey!
My first comment is about the UI presentation of the selection lists. I'm not sure how but there has to be a better way to present the not selected elements!?

![grafik](https://user-images.githubusercontent.com/2870104/29001676-741f412c-7a91-11e7-91fc-c72f4f5d882c.png)


> @ThomDietrich you might like this 😉

Indeed! Thanks ;)

> If there's just one floor, Items have no prefix (e.g. GF_Kitchen, FF_Bathroom etc).

Agreed

> If there's more than one floor, Sitemap generates them instead of single rooms. I'm not sure if it's right, but worth discussing.

Well if you ask for it, why not a frame per floor with the room groups?

Also ...

> Group item=GF
> Group item=FF
> Group:Number:AVG                 Temperature   "Temperature"   <temperature>   (Home)  

I've mentioned it somewhere else already, we should discuss style. A small but interesting problem with "GF" is, that most syntax highlightings will interpret this as a constant:

![grafik](https://user-images.githubusercontent.com/2870104/29001889-42a55c86-7a95-11e7-8536-77d047d2d724.png)

The same goes for "Temperature", is that the way of naming groups we want to promote? @kaikreuzer

kubawolanin commented:
> I'm not sure how but there has to be a better way to present the not selected elements!?

Yes. [I'm on it](http://monterail.github.io/vue-multiselect/#sub-multiple-select) ;) We'll be able to filter and select multiple items using keyboard and mouse. Styling is still on my TODO list.

> Well if you ask for it, why not a frame per floor with the room groups?

I was just thinking about it. Sure, will do!

> I've mentioned it somewhere else already, we should discuss style. A small but interesting problem with "GF" is, that most syntax highlightings will interpret this as a constant:

Yeah, we'll need to define a proper nomenclature for [the floors](https://github.com/kubawolanin/openhab-core/blob/54165a7e1d368249fc8f7bf6109a6a23b5678898/bundles/org.openhab.ui.dashboard/web/js/home-designer.js#L12).

Thanks for your input!

kubawolanin commented:
> Need to work on the order of rooms and devices selection - right now it feels random as I'm simply pushing new fields to the array.
If possible you could already design it that way, that later icons can be added for each room?

I was referring to the form's order, not the result.
Regarding the icons - yeah, I can add an option same way I did with channels.

> Besides, I think there should be an input field for custom room names (e.g. I'd like to create "StephansRoom" or "PianoRoom").

That'll require some effort - hopefully someone will contribute 😉 For the MVP I'm going to delivier a finite number of rooms/devices.

kubawolanin commented:
OK, after some long hours I've come to a point that I've decided I'm gonna create a separate bundle for this tool.
It'll be available *from* the dashboard. That said - I'm no longer relying on jQuery and can create a simple REST connector inside my Vue app 😄 

I will need some more time to create the bundle but it'll pay off.

kaikreuzer commented:
Sounds like a good decision to me! Feel free to create the bundle in openhab-core repo nonetheless.

> Hey @kaikreuzer I've actually figured it out myself :) In the upcoming commit you'll see something like this in geolocation.js.

Could you create a separate PR for this change as it is unrelated to the generator (which now ends up in a different bundle anyhow)?

kubawolanin commented:
> Could you create a separate PR for this change as it is unrelated to the generator (which now ends up in a different bundle anyhow)?

Absolutely! I'll include the assets cleanup there as well (e.g. style files in `css` folder etc).

@kaikreuzer what's your opinion on the naming convention for the group items that @ThomDietrich mentioned?

> I've mentioned it somewhere else already, we should discuss style. A small but interesting problem with "GF" is, that most syntax highlightings will interpret this as a constant:

> The same goes for "Temperature", is that the way of naming groups we want to promote? Alternatives would be "Temperatures" or "gTemperature". I'm not saying that one or the other are better choices!
> @kaikreuzer

kaikreuzer commented:
> @kaikreuzer what's your opinion on the naming convention for the group items

I don't have a strong opinion. Prefixing groups with a "g" can be helpful, although it makes it all a bit less readable. But as the demo.items use that approach for most groups, we could go for keeping it.

kirantpatil commented:
Hi @kubawolanin and @kaikreuzer ,

Is this alternative to [Eclipse SmartHome Designer](http://docs.openhab.org/installation/designer.html)?

Will it work with ESH ?

Thanks.

kaikreuzer commented:
@kirantpatil No, it is not an editor, it is a generation.
Check https://github.com/openhab/openhab-vscode if you are looking for an alternative to the Designer.

ThomDietrich commented:
Hey Kuba 👋  is the demo still up to date? Looks good but I'm concerned about the fact, that the GUI doesn't follow the Paper UI design any longer... what's your thought on that?

Minor issue I found:

![grafik](https://user-images.githubusercontent.com/2870104/29326468-e37f4786-81eb-11e7-867c-ae8299da345e.png)

kubawolanin commented:
Hi @ThomDietrich.
Was it ever following Paper UI design? 😄 
From the beginning it was using Bootstrap. We may tweak it later, when I'll have the core feature set stable and ready. 
Maybe you'd like to help with the styling a bit? [Here's](https://github.com/kubawolanin/openhab-core/blob/home/bundles/org.openhab.ui.homebuilder/README.md#development) the instructions on how to setup the dev environment.

Thanks for reporting the layout issue - I'm aware of it. This is caused by temporary fixed width of the [affix element](https://github.com/kubawolanin/openhab-core/blob/home/bundles/org.openhab.ui.homebuilder/web/src/scss/app.scss#L152).

I'm focused on bugfixing and delivering e.g. dynamically added rooms or REST-based Items.
Lots of refactoring and cleanup but I'm getting there every day 🥇

kubawolanin commented:
@kaikreuzer i have a question wrt the Items management through the REST API.

According to [the REST docs](demo.openhab.org/doc/index.html) I can do a `PUT /items/{itemname}`.
But what if I'd like to add an array of Items simultaneously?
Is there an option that allows me to do that?

As a workaround I can iterate over objects and send hundreds of requests one by one.
Not a perfect solution, though. 🤔 

Thanks!

kaikreuzer commented:
> But what if I'd like to add an array of Items simultaneously?
> Is there an option that allows me to do that?

No, there isn't. It shouldn't be too hard to add support for that. Feel free to open an issue at ESH and we can then discuss, how the REST URL for that should exactly look like.

bkumio commented:
Hi kubawolanin,

As I am not a developer I at least wanted to help today with the German translation. However I am not familiar with github @all and immediately got confused.  Perhaps you can shed some light on this.

If I look at the commit requests here I do find two language files  called
/web/i18n/pl-PL.json 
/web/i18n/en-UK.json
However I do not find a "add files" on this page. 

Then I thought ok, I most likely need to commit a new file to your generator-openhab project directly on your homepage https://github.com/kubawolanin/generator-openhab instead? 
However the commits on this page do have 72 lines and the ones on your projects do have only about 65lines and are even named differently. Just pl.json or en.json.... ?

Am I the right location for adding the German translation? me: confused :-)

kubawolanin commented:
Hey @bkumio,

I'd love to see your contribution in the project!
The reason I was not asking for help until now was that the project is in a heavy development phase right now. 
As you've pointed earlier, there might be a few rooms and devices that are worth adding.
I just don't have enough time to work on these things in parallel :-)
The JSON file structure for labels is likely to change and I simply want to avoid translation files that are incomplete.

If you'd like to get your hands dirty anyway, here's the folder containing current language files:
[web/i18n](https://github.com/kubawolanin/openhab-core/tree/home/bundles/org.openhab.ui.homebuilder/web/i18n)

And here you can find the languages/rooms/devices definitions: [definitions.js](https://github.com/kubawolanin/openhab-core/blob/home/bundles/org.openhab.ui.homebuilder/web/src/definitions.js)

Feel free to modify these files. Just remember to change `en-US.json` file as well if you're about to add/remove the items.

The easiest way for you to cooperate on this is simply by forking [my repository](https://github.com/kubawolanin/openhab-core/tree/home), cloning your fork, and then adding/modifying files on your local drive. 
`bundles/org.openhab.ui.homebuilder/web/` is the path containing homebuilder sources.
When you're done with your changes, simply `commit` them and `push` to your forked repository. Then [create a Pull Request against my repository](https://github.com/kubawolanin/openhab-core/pulls). More info on Git can be found [here](https://help.github.com/). Good luck!

kaikreuzer commented:
Hey @kubawolanin! What is the current status of this? Do you think we are getting close to some initial version that could be reviewed and merged (while being further improved subsequently)?

kubawolanin commented:
I've been busy with my other assignments lately, but I'm still on track with this PR.
I have a few things left for MVP:
- Working i18n for all Items
- (defect) Removing rooms from the floors doesn't work
- Adding multiple items within one request via REST API
- (defect) Styling bug mentioned by @ThomDietrich in his previous post
- Hide the model JSON output (left bottom side of the screen)
- Updated README

Things I'll work after I sort out these above:
- Custom rooms
- (optional) Syntax highlighting
- Align the styling with https://github.com/openhab/openhab-docs/pull/494
- Saving actual files within file system - I think I'll need someone with Java skill there. I've already implemented quick Copy button so there's workaround.

You can already play around with [the demo site](http://kubawolanin.com/home-test/) and generate yourself some items and whatnot :-) I didn't upload the icons though.

@bkumio I still remember about you! Will ping you when the project is ready for the translation files coming in :-)

ThomDietrich commented:
Hey @kubawolanin, took the chance to check out the latest demo...

* In my personal opinion, I think that questions for form elements are rather unusual. I think the whole HomeBuilder would be easier to use if it read "Home Setup Name" instead of "How do you want to call your home?" or "Configuration Storage" instead of "How would you like to store the Items?"

* "Directly in the JSON database": Directly sounds opinionated and JSON is a technical term the end user shouldn't need to know. Do you agree? How about "Textual Configuration Files" vs "Internal Database"?
(btw. be aware that you are not only generating *.items files)

* A few elements have a red star at the end of the label. I guess this is supposed to indicate mandatory inputs but I'd rather use the validation options your framework probably provides.

* In he rooms and devices selection there is a label "Press enter to select" on the right. I'd remove it. A keyboard user probably knows what to do, a mouse or touch user is not addressed to begin with.

* "Please select the smart devices for: Attic (GroundFloor)"
  "Please select the smart devices for: Bedroom (GroundFloor)"
  and so on are not easy to read. As said before I'd suggest to reduce the labels. How about Headlines or Frames for "Floors", "Rooms" and "Devices" with the Labels reduced to "Attic"?

* `{ channel="" }` - Why the extra spaces? None of the other parts (icon, groups, etc) have them.

* Where and how will the Home Builder be visible? Will it be part of the Setup process or/and one of the selectable UIs next to PaperUI, BasicUI etc. after the initial package selection? Imho there should be some more guidance directly on that page. You've mentioned a README you need to enhance. Ideally the Home Builder should be self explanatory. For that to be true, I'd say there should be a short introduction and a short conclusion about What the Builder offers and what to do next (without going into specific details). Furthermore every form element should have a tooltip with an explanation (e.g. "Include Tags").

Great work so far, thanks for taking my comments into consideration. Let me know what you think. Best!

kubawolanin commented:
@bkumio I'm coming back with the answers finally! :-)

> The first questions should be translated too, or? Your english file has got the questions but the PL language file doesn’t?

For now I'll keep them English only. Further development may include i18n of the form itself.

> I realized that for some words for singular and plural do have the same word in German. Does this create a problem if we generate two items with the potential same name?

I ditched the plural forms in favor of `g` prefix for groups. No idea about approach for German translation, though.

> Any visualisation problems with using the German Umlauts "ä/ö/ü/ ?

Shouldn't be any :-) Just remember to provide JSON file encoded in UTF-8 and it should be OK.

> Shutters I think we need another one called "Blinds" as these are two different things, or?

You got it!

> Perhaps worth adding a “kids bath(room)” ?
Perhaps worth adding a “garden”
Perhaps worth adding a “penthouse”
…. The are even more 😊

I've limited the number of rooms a bit. They're now aligned with Google Home app (just for sake of clarity) + I've added some additional ones.

The good news is that I've finally managed to make custom rooms working! Well, *almost*
You can add a custom room by simply typing its label in the floor's field.
Then you need to remove it and select it from the list again. It's a bug I'm aware of.

> Plus another feature request would be to create Items already prepared for KNX syntax . In the sense of we could have the items already with a "{knx = "0/0/0"} for a switch or "{ knx = "0/0/0+<1/1/1" }" incl. the status input GA. Devil is in the detail for the shutters and other stuff

I don't want this tool to be binding-oriented. Therefore, you'll be able to append `{channel=""}` to the textual Items config. Then you're free to fine-tune it and replace `channel` string in the file to whatever you prefer :-)

---

@ThomDietrich many thanks for your kind words and suggestions!

> In my personal opinion, I think that questions for form elements are rather unusual. 

Renamed them as suggested.

> "Directly in the JSON database": Directly sounds opinionated and JSON is a technical term the end user shouldn't need to know. Do you agree? 

Absolutely! Changed that as well

> A few elements have a red star at the end of the label. I guess this is supposed to indicate mandatory inputs but I'd rather use the validation options your framework probably provides.

Yep - I've removed them.

> In he rooms and devices selection there is a label "Press enter to select" on the right. I'd remove it.

You got it mate!

> "Please select the smart devices for: Attic (GroundFloor)"
...
How about Headlines or Frames for "Floors", "Rooms" and "Devices" with the Labels reduced to "Attic"?

Brilliant idea! I was actually struggling with the right presentation of it. 
Implemented!

> { channel="" } - Why the extra spaces? None of the other parts (icon, groups, etc) have them.

Well, it looks kinda cleaner to me with the spaces. But looking [at the docs](http://docs.openhab.org/configuration/items.html#item-definition-and-syntax) I see that there are no spaces as well so I've removed it here too.

> Where and how will the Home Builder be visible? Will it be part of the Setup process or/and one of the selectable UIs next to PaperUI, BasicUI etc. after the initial package selection? 

That's correct - it's a new UI accessible from the dashboard (`/start/index`)

> Imho there should be some more guidance directly on that page. You've mentioned a README you need to enhance. Ideally the Home Builder should be self explanatory. For that to be true, I'd say there should be a short introduction and a short conclusion about What the Builder offers and what to do next (without going into specific details). Furthermore every form element should have a tooltip with an explanation (e.g. "Include Tags").

I fully agree. Although this is something we could tweak/enhance on the go. 

Good news is that I've finally finished `i18n` support for the tool!
Now let's fight this pesky [styling bug](https://github.com/openhab/openhab-core/pull/172#issuecomment-322525741) :-)

kubawolanin commented:
> Happy to help with German/French translations and testing 😊

@bkumio would you like to include German/French translation now?
I'll make it simpler for you - just [create a new Gist](https://gist.github.com/) with `fr-FR.json` and `de-DE.json` files based on [this file in my repository](https://raw.githubusercontent.com/kubawolanin/openhab-core/home/bundles/org.openhab.ui.homebuilder/web/i18n/en-UK.json).
There's a "Add file" button on bottom left side of the screen.

When you're ready, I'll include the translations. Just please comment here that you agree for me to include your sign-off within this PR. Read more about the formal process [here](https://github.com/openhab/openhab-core/blob/master/CONTRIBUTING.md#sign-your-work).

@ThomDietrich and others - I've updated [my demo site](http://kubawolanin.com/home-test/) with the latest code. Please make sure you've cleaned the cache (`Ctrl + Shift + R` in Chrome should do) before testing.

@kaikreuzer @htreu I think we can go ahead with the review :-)

Cheers!

bkumio commented:
yes , can do the translations tomorrow I think....

cheers
   Mario

bkumio commented:
Hi kubawolanin,

translations  (German / French) you will find on this gist : 
https://gist.github.com/bkumio/6077b0eff8021b433e7e22de3e26a72a

hopefully this is the format you were looking for.

Mario

ThomDietrich commented:
Hey @bkumio, may I add a few suggestions?

```json
"Attic": "Dachboden",
"Bathroom": "Bad/Badezimmer",
"Entryway": "Zufahrt",
"MasterBedroom": "Hauptschlafzimmer",
"Outside": "Draußen",
"Fan": "Ventilator",
```

bkumio commented:
...allways good to have a 2nd pair of eyes. Cant see the wood for the trees anymore :-)


----- file is updated with the comments. ----

"Attic": "Dachboden", ---> hm good question. I translated the Loft as "Speicher" and put this as Attika as that´s something which is "outside" whereas the Loft is inside e.g. Dachboden. Or AmI wrong here ? 

"Bathroom": "Bad/Badezimmer",  --->Makes sense, missed it!

"Entryway": "Zufahrt",  ---> makes sense. Never had so much money that I could afford one

"MasterBedroom": "Hauptschlafzimmer",  --> makes sense

"Outside": "Draußen",   ---> well if you like German special signs. Aber doch nicht mehr nach der Rechtschreibreform, oder ? I left it as is but if more people think this should change we change 

"Fan": "Ventilator", ---> makes sense ,mixed it up with my helios central air exchange which is more a Lüftung....

ThomDietrich commented:
For me the Attic is "Dachboden", the word "Speicher" is uncommon in my area (Thüringen). Truthfully I have to say that I've never heard or used the word "Attika" 🙉 

There is a big difference between attic and loft, let's ask google:
Attic: https://www.google.de/search?q=what+is+an+attic&source=lnms&tbm=isch
Loft: https://www.google.de/search?q=what+is+a+loft&source=lnms&tbm=isch

draußen: https://de.wiktionary.org/wiki/draussen

Fan: well it depends I guess. A fan is first and foremost a "Ventilator" (standing or on the ceiling). It may of course be part of a fixed house-wide "Lüftung" but funnily that is called ventilation in english.

bkumio commented:
Ha amazing....  I think I really got more influenced by the area I grew up than I thought.
Perhaps I need to create a  bd-DE (Badische) version of the translation :-). 

Dachboden in Basel is  btw. "E-Strich" which is just a concrete floor for me in Freiburg. You are probably right and caught me on "Speicher" which is indeed "Badisch.  "Attika" is as well South German for the surrounding wall on a flat rooftop. http://www.differencebetween.com/difference-between-loft-and-vs-attic/  .Bottom line, I am happy with "Dachboden" .
But I have changed Loft back to "Loft" as this was Dachboden too.

Ok ok , "Draußen" it is. As I work for 20 years now in Basel/CH and as well have a swiss German keyboard  @home I don´t have that "ß" on my keyboard and don’t miss it.

See you learn something everyday

kubawolanin commented:
@bkumio @ThomDietrich I've added the translation files to the app -> [http://kubawolanin.com/home-test/]()

Thank you for your work! :de: :fr: 

There's one bug I'm aware of - when you switch languages, the "Ground Floor" field's label doesn't change. Same with the selected pills you've chosen before language switch. It doesn't affect the end result. No ETA for the fix.

bkumio commented:
Hi kubawolanin,

almost there, one more minor thing I have realized is that some rooms do not have an icon associated with it , perhaps you can add this on the todo.

ah, can you read the german file again pls. I have made some final changes acroding to the suggestions from Thom

bodiroga commented:
Hi @kubawolanin!

I've translated the strings to Spanish, I hope you can add them before the final version is merged ;-) Here you have my gist link:

https://gist.github.com/bodiroga/5acabc10b5e654d23b54bac81557ea58

Great work with HomeBuilder, it's an awesome tool!

Many thanks for all your effort and best regards,

Aitor

ThomDietrich commented:
If it is desirable, I can get an italian translation from a colleague, probably within the next 12h

ThomDietrich commented:
Or less. Here we go: https://gist.github.com/ThomDietrich/4ca0b0790fb5af192d4cadcace0dd49d

hakan42 commented on bundles/org.openhab.ui.homebuilder/pom.xml:
```diff
@@ -0,0 +1,19 @@
+
+	<parent>
+	    <groupId>org.openhab.core</groupId>
+    	<artifactId>pom-bundles</artifactId>
```
some tab and space confusion in the pom 😄

ThomDietrich replied:
My favorite comment for such a case: https://www.emacswiki.org/pics/static/TabsSpacesBoth.png

maggu2810 replied:
> My favorite comment for such a case: https://www.emacswiki.org/pics/static/TabsSpacesBoth.png

:+1: :rofl:

hakan42 commented on bundles/org.openhab.ui.homebuilder/src/main/java/org/openhab/ui/homebuilder/internal/HomeBuilderDashboardTile.java:
```diff
@@ -0,0 +1,73 @@
+
+    public static final String HOMEBUILDER_ALIAS = "/homebuilder";
+
+    private final Logger Logger = LoggerFactory.getLogger(HomeBuilderDashboardTile.class);
```
the logger object should start with a lowercase "l" 😄

hakan42 commented on bundles/org.openhab.ui.homebuilder/src/main/java/org/openhab/ui/homebuilder/internal/HomeBuilderDashboardTile.java:
```diff
@@ -0,0 +1,73 @@
+    protected void activate() {
+        try {
+            httpService.registerResources(HOMEBUILDER_ALIAS, "web", null);
+            Logger.info("Started HomeBuilder at " + HOMEBUILDER_ALIAS);
```
Please avoid string concatenation in logging statements.

hakan42 commented on bundles/org.openhab.ui.homebuilder/web/dist/build.js:
```diff
@@ -0,0 +1,37 @@
+!function(t){function e(r){if(n[r])return n[r].exports;var i=n[r]={i:r,l:!1,exports:{}};return t[r].call(i.exports,i,i.exports,e),i.l=!0,i.exports}var n={};e.m=t,e.c=n,e.i=function(t){return t},e.d=function(t,n,r){e.o(t,n)||Object.defineProperty(t,n,{configurable:!1,enumerable:!0,get:r})},e.n=function(t){var n=t&&t.__esModule?function(){return t.default}:function(){return t};return e.d(n,"a",n),n ...
```
Are the files in "web/dist" meant to be checked in? I thought they were created during build time?

kubawolanin replied:
For now they are, but I plan to chain it with Maven like @resetnow did with BasicUI in  https://github.com/eclipse/smarthome/pull/4040

kubawolanin commented:
Hello everyone,

I've addressed comments by @hakan42. Thank you for the review!
There are also translation files already in place. Thank you @bkumio @bodiroga @ThomDietrich (and his colleague! :D) for making this tool multilingual! :smiley: 

I look forward to see other languages implemented as well!

The [demo site](http://kubawolanin.com/home-test/) is now updated with the latest code.
One important change I'd like to highlight is the sitemap generator.
For now the device groups are generated inside the `Text` node.

```java
    ...

    Frame {
        Text label="Window" {
            Default item=GF_Backyard_Window label="Backyard"
        }

        Text label="Motion Sensor" {
            Default item=GF_Backyard_Motion label="Backyard"
            Default item=GF_Bathroom_Motion label="Bathroom"
            Default item=GF_Boiler_Motion label="Boiler Room"
        }

        Text label="Roller Shutter" {
            Default item=GF_Backyard_Shutter label="Backyard"
        }

        Text label="Power Outlet" {
            Default item=GF_Backyard_Power label="Backyard"
        }
    }
```

Before, it was impossible to differentiate items within these sections - they were all labeled e,g, "Motion Sensor". 

As always, feel free to play around with the tool and let me know what you think!

@ThomDietrich would you mind taking a look at it again from the usability standpoint?
I'm sure there are tons of things to be tweaked/improved.
I'd like to, however, address them sequentially. 

Thank you all for your help!

bkumio commented:
Hi kubawolanin,

just some UI/functionality whish on the items. I usually have more than one window or light or blinds in a room. I think It would be great If I can add any item like the "window" or the blinds more than just one time and have the possibility to change the display name directly in the UI

For example
Rollershutter   FF_Kitchen_Shutter   "Shutter south"  <rollershutter>  
Rollershutter   FF_Kitchen_Shutter   "Shutter west"  <rollershutter>  
Rollershutter   FF_Kitchen_Shutter   "Shutter all"  <rollershutter> 

That would allow me to generate the sitemap/Item structure ,much more tailored to my needs. But yes I can create the structure and then copy/paste the text afterwards too. Depends on if we see this tool as a basic structure generator or a "setup wizard" ....

Just a thought
  Mario

ThomDietrich commented:
Definitely a "setup wizard". However, the option to be able to add more than one Window indeed sounds useful!

@kubawolanin I'll add my own comments soon.

kubawolanin commented:
@ThomDietrich @kaikreuzer :ping_pong: :-)

claudioita commented:
Possible to put:
HUE lamps (not only lights but also colours
Chromecast
and others

ThomDietrich commented:
Hey Kuba 🏓 

- "Please select your language confuses as it doesn't change the UI elements and is the first form element on he page. If it's only supposed to influence the language of the generated Items, maybe rename it in this direction and move it down to near bottom?

- When using tags "Motion Sensor" is tagged as switchable. Intended?

- I'm still missing the explanatory text elements, especially the one explaining what to do with the generated result. A new user using the wizard needs a hint on what to do next. That could be as easy as referencing a README on docs.openhab.org. A short description would be better...

Besides these I feel like you've reached a great result! 🎉 🍻

kubawolanin commented on bundles/org.openhab.ui.homebuilder/web/src/sitemap.js:
```diff
@@ -0,0 +1,80 @@
+
+        if (model.floorsCount > 1) {
+            let icon = model.itemsIcons && floor.icon ? '" icon="' + floor.icon : '';
+            lines.push('    Frame label="' + (floor.name || floor.value) + icon + '" {');
```
`s.pad(' ', 4)`

kubawolanin commented on bundles/org.openhab.ui.homebuilder/web/src/sitemap.js:
```diff
@@ -0,0 +1,80 @@
+            lines.push('    Frame label="' + (floor.name || floor.value) + icon + '" {');
+            floorPrefix = floor.abbr + '_';
+        } else {
+            lines.push('    Frame {');
```
`s.pad(' ', 4)`

kubawolanin commented on bundles/org.openhab.ui.homebuilder/web/src/sitemap.js:
```diff
@@ -0,0 +1,80 @@
+
+        if (floor && floor.value && !_.isUndefined(model[floor.value])) {
+            model[floor.value].forEach(function(room) {
+                lines.push('        Group item=' + floorPrefix + room.value);
```
`s.pad(' ', 8)`

kubawolanin commented on bundles/org.openhab.ui.homebuilder/web/src/sitemap.js:
```diff
@@ -0,0 +1,80 @@
+                lines.push('        Group item=' + floorPrefix + room.value);
+            });
+        }
+        lines.push('    }');
```
`s.pad(' ', 4)`

kubawolanin commented on bundles/org.openhab.ui.homebuilder/web/src/textItems.js:
```diff
@@ -0,0 +1,167 @@
+
+/**
+ * Generates a "channel" string for the device item
+ * e.g. `{ channel ="" }`
```
`{channel=""}`

108adams commented on bundles/org.openhab.ui.homebuilder/web/src/AppForm.vue:
```diff
@@ -0,0 +1,206 @@
+  </div>
+</template>
+
+<style lang="scss" src="./scss/app.scss">
```
dead code

kubawolanin replied:
Note that `src` parameter is used here. The empty line between `style` tag though is a result of JS Beautifier's formatter being funky in `*.vue` files ;-)

108adams commented on bundles/org.openhab.ui.homebuilder/web/src/AppForm.vue:
```diff
@@ -0,0 +1,206 @@
+    /**
+     * Generates textual Items
+     */
+    generateItems: generateItems,
```
ES6 syntax

108adams commented on bundles/org.openhab.ui.homebuilder/web/src/AppForm.vue:
```diff
@@ -0,0 +1,206 @@
+      this.$http
+        .get('services/org.eclipse.smarthome.core.i18nprovider/config')
+        .then(response => {
+          let body = response.body;
```
use const in place of let. Immutable by design

108adams commented on bundles/org.openhab.ui.homebuilder/web/src/AppForm.vue:
```diff
@@ -0,0 +1,206 @@
+     */
+    fetchTranslations: function(language) {
+      let root = window.location.href.replace('/index.html', '');
+      this.$http.get(root + '/i18n/' + language + '.json').then(response => {
```
formatting, stay consistent, `.then` in next line

108adams commented on bundles/org.openhab.ui.homebuilder/web/src/AppForm.vue:
```diff
@@ -0,0 +1,206 @@
+        let roomsModel = _.chain(this.$data.model)
+          .pickBy((value, key) => {
+            return _.endsWith(key, DEVICES_SUFFIX);
+          })
```
```
.pickBy((value, key) => _.endsWith(key, DEVICES_SUFFIX))
```

108adams commented on bundles/org.openhab.ui.homebuilder/web/src/AppForm.vue:
```diff
@@ -0,0 +1,206 @@
+          })
+          .value();
+
+        floors.forEach(function(item) {
```
arrow function

108adams commented on bundles/org.openhab.ui.homebuilder/web/src/Collapse.vue:
```diff
@@ -0,0 +1,80 @@
+            copyText: 'Copy',
+            createText: 'Create Items'
+        }
+    },
```
data: () => ({
            copyText: 'Copy',
            createText: 'Create Items'
        }),

108adams commented on bundles/org.openhab.ui.homebuilder/web/src/Collapse.vue:
```diff
@@ -0,0 +1,80 @@
+                this.$http
+                    .put('items', { body: content })
+                    .then(response => {
+                        if (response.status === 200) {
```
else?

kubawolanin replied:
Done.

108adams commented on bundles/org.openhab.ui.homebuilder/web/src/Collapse.vue:
```diff
@@ -0,0 +1,80 @@
+                                this.createText = 'Create Items';
+                            }, 500);
+
+                            console.log("Items created successfully!");
```
remove

108adams commented on bundles/org.openhab.ui.homebuilder/web/src/habpanel.js:
```diff
@@ -0,0 +1,81 @@
+        var floor = floors[i];
+
+        if (floor && floor.value && !_.isUndefined(model[floor.value])) {
+            model[floor.value].forEach(function(room) {
```
very high complexity, refactor

kubawolanin replied:
Done. Thanks!

108adams commented on bundles/org.openhab.ui.homebuilder/web/src/habpanel.js:
```diff
@@ -0,0 +1,81 @@
+                        });
+                    }
+
+                    widgets.push(widget);
```
reduce pattern

108adams commented on bundles/org.openhab.ui.homebuilder/web/src/formSchema.js:
```diff
@@ -0,0 +1,275 @@
+ * @param {Object} field 
+ */
+function roomsChanged(model, newVal, oldVal, field) {
+    let devicesFields = this.$options.parent.groups[3].fields;
```
magic number

108adams commented on bundles/org.openhab.ui.homebuilder/web/src/formSchema.js:
```diff
@@ -0,0 +1,275 @@
+    let devicesFields = this.$options.parent.groups[3].fields;
+    let floor = field.model;
+    let oldList = oldVal ? oldVal.map((room) => room.value) : [];
+    let newList = newVal.map((room) => room.value);
```
https://stackoverflow.com/questions/35136306/what-happened-to-lodash-pluck

```
let newList = _.map(newVal, 'value');
```

kubawolanin commented on bundles/org.openhab.ui.homebuilder/web/src/textItems.js:
```diff
@@ -0,0 +1,167 @@
+        .map((line) => line.slice(2).trimRight())
+        .join('\n');
+
+    // Add device groups, e.g. "gTemperature"
```
Old comment

kubawolanin commented on bundles/org.openhab.ui.homebuilder/web/src/habpanel.js:
```diff
@@ -0,0 +1,81 @@
+        let device = _.find(devices, { value: dev });
+        return {
+            id: device.value,
+            name: device.name || device.value, // TODO: plural?
```
Remove TODO

kubawolanin commented on bundles/org.openhab.ui.homebuilder/web/src/textItems.js:
```diff
@@ -0,0 +1,167 @@
+            }
+            break;
+        case 'device':
+            // type = '    ' + type; // TODO: Grouping devices in rooms
```
Remove this case altogether.

108adams commented on bundles/org.openhab.ui.homebuilder/web/src/formSchema.js:
```diff
@@ -0,0 +1,275 @@
+    styleClasses: 'floors-number',
+    values: [1, 2, 3, 4, 5],
+    onChanged: function(model, newVal, oldVal, field) {
+        let roomsFields = this.$options.parent.groups[2].fields;
```
magic number

kubawolanin commented on bundles/org.openhab.ui.homebuilder/web/src/restItems.js:
```diff
@@ -0,0 +1,210 @@
+                ...addDevices(room, model, floorPrefix, roomDevices)
+            ];
+        });
+
```
Remove empty line

108adams commented on bundles/org.openhab.ui.homebuilder/web/src/formSchema.js:
```diff
@@ -0,0 +1,275 @@
+    label: 'Number of floors',
+    styleClasses: 'floors-number',
+    values: [1, 2, 3, 4, 5],
+    onChanged: function(model, newVal, oldVal, field) {
```
pass by reference, refactor complexity

kubawolanin replied:
[it's a part of API](https://icebob.gitbooks.io/vueformgenerator/content/fields/field_properties.html#common-events-of-field), so not much I can do here.

kubawolanin commented on bundles/org.openhab.ui.homebuilder/web/src/restItems.js:
```diff
@@ -0,0 +1,210 @@
+        case 'Switch':
+        case 'Dimmer':
+        case 'Color':
+            tags.push('Switchable');
```
`Motion Sensor` switch shouldn't be Switchable. I need to revisit this switch.

108adams commented on bundles/org.openhab.ui.homebuilder/web/src/restItems.js:
```diff
@@ -0,0 +1,210 @@
+export function addRooms(floor, model) {
+    let items = [];
+
+    if (floor && floor.value && !_.isUndefined(model[floor.value])) {
```
if _ lodash is available globally, please review other code and change `typeof` accordingly.

kubawolanin replied:
It is everywhere now except `Collapse` and custom `Multiselect` components

108adams commented on bundles/org.openhab.ui.homebuilder/web/src/restItems.js:
```diff
@@ -0,0 +1,210 @@
+}
+
+export function addDevices(room, model, floorPrefix, roomDevices) {
+    let items = [];
```
use reverted if and multiple returns

```javascript
let deviceCollection = model[roomDevices] || [];

if (!room.value || _.isEmpty(deviceCollection)) {
    return [];
}
 
return deviceCollection.map(device => ({
               type: _.first(device.type.split(':')),
               name: floorPrefix + room.value + '_' + device.value,
               label: device.name || device.value,
               category: model.itemsIcons ? device.icon : '',
               groupNames: [
                   floorPrefix + room.value,
                   GROUP_PREFIX + device.value
               ],
               tags: addTags(device, model),
               entryType: 'device'
       })
)
 ```

108adams commented on bundles/org.openhab.ui.homebuilder/web/src/restItems.js:
```diff
@@ -0,0 +1,210 @@
+ */
+export function getChosenDevices(model) {
+    return _.chain(model)
+        .pickBy((value, key) => {
```
remove return

kubawolanin commented:
@108adams thank you for thorough code review! I needed that :-) I believe I've fixed most of your remarks.

kaikreuzer commented on bundles/org.openhab.ui.homebuilder/src/main/java/org/openhab/ui/homebuilder/internal/HomeBuilderDashboardTile.java:
```diff
@@ -0,0 +1,73 @@
+            httpService.registerResources(HOMEBUILDER_ALIAS, "web", null);
+            logger.info("Started HomeBuilder at {}", HOMEBUILDER_ALIAS);
+        } catch (NamespaceException e) {
+            logger.error("Error during HomeBuilder startup: {}", e.getMessage());
```
Home Builder

kaikreuzer commented on bundles/org.openhab.ui.homebuilder/src/main/java/org/openhab/ui/homebuilder/internal/HomeBuilderDashboardTile.java:
```diff
@@ -0,0 +1,73 @@
+
+    @Override
+    public String getImageUrl() {
+        return "../homebuilder/tile.png";
```
I'd suggest to change the tile.png to be a screenshot of the tool as this is what we have for all other tiles as well. Introducing a new type of logo can always be tricky, so better refrain from it.

kubawolanin replied:
Done!

ThomDietrich commented on bundles/org.openhab.ui.homebuilder/README.md:
```diff
@@ -0,0 +1,43 @@
+# Home Builder
```
Good point!
A drawback of this approach is that the README is presented on GitHub by default and can rightfully be dev-oriented. For that reason I think a valid alternative would be `docs/USAGE.md` or similar for docs.openhab.org. Compare: https://github.com/openhab/openhab-docs/blob/gh-pages/pom.xml#L419-L431

kaikreuzer replied:
Ah, how clever of you :-)
I wasn't yet aware of that mechnism in mycroft. Yes, makes sense, so I correct my request to @kubawolanin: Please add a USAGE.md instead 😎

kubawolanin replied:
Done!

kubawolanin commented on bundles/org.openhab.ui.homebuilder/web/src/definitions.js:
```diff
@@ -0,0 +1,87 @@
+  { value: 'Upstairs', icon: 'firstfloor' }
+];
+
+export var devices = [
```
"objects" seem to be too vague, especially in the JavaScript context. I could go with "smartObjects" or "smartThings" or maybe just "things"? Would like to avoid confusion too.

kaikreuzer replied:
"Things" is already the word for the physical side of openHAB, definitely not the right term for items.
"smart" is also not correct as your windows is not smart.
Check out https://youtu.be/R1HVYWP51E0. I tentively used "BuildingObjects" there, but that is only valid for windows & doors and alike. But if you want to cover also furniture, lights, tvs, etc. "Object" is the only reasonable term, really.

kubawolanin commented on bundles/org.openhab.ui.homebuilder/web/src/definitions.js:
```diff
@@ -0,0 +1,87 @@
+  { value: 'Humidity', icon: 'humidity', type: 'Number:AVG', unit: '[%d %%]' },
+];
+
+export const DEVICES_SUFFIX = '_devices';
```
See my [previous comment](https://github.com/openhab/openhab-core/pull/172#discussion_r146040327) on that

kaikreuzer commented:
@kubawolanin Thanks for addressing my comments so quickly (and sorry for coming back so late...). Is the code in state that you would suggest it to be merged as an initial version (I am asking since there are unchecked tasks listed above)?

kubawolanin commented:
Hey @kaikreuzer! The unchecked tasks are here for information purposes and they can be easily moved to Issues as soon as the branch is merged.
I have one uncommitted refactoring in `sitemap.js` that I'd like to push today evening. After that I think we're ready to go :-)

martinvw commented on bundles/org.openhab.ui.homebuilder/src/main/java/org/openhab/ui/homebuilder/internal/HomeBuilderDashboardTile.java:
```diff
@@ -0,0 +1,73 @@
+/**
+ * Copyright (c) 2015-2017 by the respective copyright holders.
```
I don't know whether the license goal works on this repo but you will find out :-)

kubawolanin replied:
Hmm, seems that [DashboardTile.java](https://github.com/kubawolanin/openhab-core/blob/master/bundles/org.openhab.ui.dashboard/src/main/java/org/openhab/ui/dashboard/DashboardTile.java#L2) has the same copyright header.

I've run `mvn license:format` *after* I changed `2015` to `2010` and I realized that it changed it back ;)

martinvw commented on bundles/org.openhab.ui.homebuilder/web/src/restItems.js:
```diff
@@ -0,0 +1,213 @@
+    }
+
+    items = [
+        ...items,
```
What does this syntax do?

kubawolanin replied:
It's JavaScript's [spread operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_operator).

martinvw replied:
Goal achieved I learned something today 😉

kaikreuzer commented:
> I have one uncommitted refactoring in sitemap.js that I'd like to push today evening. After that I think we're ready to go :-)

I didn't find any commits after this comment and @martinvw's partial review, so I assume this PR is still on your side and not waiting to be merged. If this is a false impression, please scream ;-)

kubawolanin commented:
That's correct @kaikreuzer, I've focused my forces on VSCode lately but I'm back to the game ;-)

kubawolanin commented on bundles/org.openhab.ui.homebuilder/OSGI-INF/dashboardtile.xml:
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+
+    Copyright (c) 2015-2017 by the respective copyright holders.
```
See [my previous comment](https://github.com/openhab/openhab-core/pull/172#discussion_r153050002)

kubawolanin commented on bundles/org.openhab.ui.homebuilder/.eslintrc.json:
```diff
@@ -0,0 +1,40 @@
+        "no-this-before-super": 1,
+        "constructor-super": 1
+      }
+}
```
Sure!

kubawolanin commented on bundles/org.openhab.ui.homebuilder/build.properties:
```diff
@@ -0,0 +1,14 @@
+               web/
+output = target/classes/
+source.. = src/main/java/
+bin.excludes = web/node_modules/,\
```
I expect the number of included files to grow anyway, so let's keep it as it is.

kubawolanin commented:
@kaikreuzer @martinvw I've addressed most of the comments.
I'm not sure how to apply DS annotations properly so hopefully it'll come in a new PR afterwards.

kaikreuzer commented on bundles/org.openhab.ui.homebuilder/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,22 @@
+ org.osgi.framework,
+ org.osgi.service.cm,
+ org.osgi.service.http,
+ org.osgi.service.component.annotations;resolution:=optional,
```
this import should not be required
